creator-node/src/ipfsClient.js

```diff
@@ -27,78 +34,183 @@ async function logIpfsPeerIds () {
-  }
 
-  const ipfsDaemonHashLogStr = `ipfsDaemonHash=${ipfsDaemonHashes} isSameHash=${onlyHashes === ipfsDaemonHashes}`
-  logger.info(`[ipfsClient - ipfsAddWrapper()] onlyHash=${onlyHashes} ${ipfsDaemonHashes ? ipfsDaemonHashLogStr : ''}`)
+    const startIpfsLatestAdd = hrtime.bigint()
+    for await (const ipfsLatestAddWithDaemonResp of ipfsLatest.add(buffer, ipfsConfig)) {
+      ipfsDaemonHash = `${ipfsLatestAddWithDaemonResp.cid}`
+    }
+    const durationIpfsLatestAddMs = (hrtime.bigint() - startIpfsLatestAdd) / BigInt(1000000) // convert ns -> ms
 
-  // If content was added to ipfs daemon, prioritize using that hash response
-  return ipfsDaemonHashes || onlyHashes
+    logger.info(`[ipfsClient - ipfsAddNonImages()] onlyHash=${onlyHash} onlyHashDuration=${durationOnlyHashMs}ms ipfsDaemonHash=${ipfsDaemonHash} ipfsDaemonHashDuration=${durationIpfsLatestAddMs}ms isSameHash=${onlyHash === ipfsDaemonHash}`)
+    return ipfsDaemonHash
+  } catch (e) {
+    logger.warn(`[ipfsClient - ipfsAddNonImages()] Could not add content to ipfs. Defaulting to onlyHash=${onlyHash}: ${e.toString()}\n${e.stack}`)
+    return onlyHash
+  }
 }
 
 /**
- * Wrapper to ipfs.add() -- This wrapper allows for ipfs calls to be made async or synchronous.
- * @param {Object} ipfs ipfs library -- can be two different versions (see ipfsClient.js)
- * @param {string|string[]} srcPath
+ * Wrapper for ipfs.add() with flag to exclusively generate the CID using the only hash logic, or to also add content to the ipfs daemon.
+ * Used for adding images to also generate dir CIDs.
+ * @param {Object} ipfsLatest ipfs instance (should be v43.0.1)
+ * @param {Object[]} content an Object[] with the structure [{ path: string, content: buffer }, ...]
  * @param {Object?} ipfsConfig ipfs add config options
  * @param {Object?} logContext
- * @returns {string|string[]|Object|Object[]} hashes from content addressing fn, or ipfs daemon responses
+ * @param {boolean?} enableIPFSAdd flag to add content to ipfs daemon
+ * @returns {Object[]} only hash responses or ipfs daemon responses with the structure [{path: <string>, cid: <string>, size: <number>}]
  */
-async function ipfsAddFromFsWrapper (ipfs, srcPath, ipfsConfig = {}, logContext = {}) {
+async function ipfsAddImages (ipfsLatest, content, ipfsConfig = {}, logContext = {}, enableIPFSAdd = false) {
   const logger = genericLogger.child(logContext)
 
-  const stream = fs.createReadStream(srcPath)
-  const onlyHash = await ipfsHashOf(stream)
-  if (ENABLE_ASYNC_IPFS_ADD) {
-    logger.info(`[ipfsClient - ipfsAddFromFsWrapper()] onlyHash=${onlyHash}`)
-    ipfs.addFromFs(srcPath, ipfsConfig) // Do not await it
-    return onlyHash
+  const startOnlyHash = hrtime.bigint()
+  const ipfsAddWithoutDaemonResp = await ipfsOnlyHashImages(content)
+  const durationOnlyHashMs = (hrtime.bigint() - startOnlyHash) / BigInt(1000000) // convert ns -> ms
+
+  const ipfsAddWithoutDaemonRespStr = JSON.stringify(sortKeys(ipfsAddWithoutDaemonResp))
```

Comment:
if we don't return the sorted then why sort at all? we have no error handling for order divergence so either 1) we assume this never happens and don't sort or 2) assume this may happen, still don't sort, but throw error on order divergence
wdyt?

Reply:
the sort function actually only sorts the inner object keys and no positional sorting occurs. for example:
```
> const in2 = [
...      {
...        "cid": "QmSRyKvnXwoxPZ9UxqxXPR8NXjcPYBEf1qbNrXyo5USqLL",
...        "path": "blob/150x150.jpg",
...        "size": 3091
...      },
...      {
...        "cid": "QmQQMV9TXxRmDKafZiRvMVkqUNtUu9WGAfukUBS1yCk2ht",
...        "path": "blob/480x480.jpg",
...        "size": 20743
...      },
...      {
...        "cid": "Qmd8cDdDGcWVaLEoJPVFtkKhYMqvHXZTvXcisYjubFxv1F",
...        "path": "blob/1000x1000.jpg",
...        "size": 72621
...      },
...      {
...        "cid": "QmaYCPUH8G14yxetsMgW5J5tpTqPaTp3HMd3EAyffZKSvm",
...        "path": "blob/original.jpg",
...        "size": 185844
...      },
...      {
...        "path": "zzzzzzz",
...        "size": 1,
...        "cid": "zzzzz",
...      }
...     ]
undefined
> sortKeys(in2)
[
  {
    cid: 'QmSRyKvnXwoxPZ9UxqxXPR8NXjcPYBEf1qbNrXyo5USqLL',
    path: 'blob/150x150.jpg',
    size: 3091
  },
  {
    cid: 'QmQQMV9TXxRmDKafZiRvMVkqUNtUu9WGAfukUBS1yCk2ht',
    path: 'blob/480x480.jpg',
    size: 20743
  },
  {
    cid: 'Qmd8cDdDGcWVaLEoJPVFtkKhYMqvHXZTvXcisYjubFxv1F',
    path: 'blob/1000x1000.jpg',
    size: 72621
  },
  {
    cid: 'QmaYCPUH8G14yxetsMgW5J5tpTqPaTp3HMd3EAyffZKSvm',
    path: 'blob/original.jpg',
    size: 185844
  },
  { cid: 'zzzzz', path: 'zzzzzzz', size: 1 }
]
```

the reason why I added the sort and then comparing the json string was bc I observed during development that the the string comparison would fail like:
```
[{"cid": "abc", "path": "some/path", "size": 123}] !== [{"path": "some/path", "size": 123, "cid": "abc"}]
```

^ these strings are clearly inequal, however, with a sort, this comparison will be equal

Reply:
then, in resizeImage.js, we only rely on position for grabbing the dir instance. this logic should always return the dir as the last index 😃

Reply:
ahh i see, i didn't realize the sortKeys method sorts keys within object, i thought it sorted an array of objects based on keys

could you make the method name more clear or add a comment?

also TIL `typeof <array> = object` lol

also i think you could prob just do a deep object-array comparison with lodash but no need to change 👍

Reply:
and thanks for the clarification 🙂

Reply:
ohhh you know, lodash is probably a better option since the intentions are more clear off the bat + functionality is the same. I think i'll just go ahead and make that change